Removed duplicate deleteBlob methods

gfyoung opened the pull request:
Title is self-explanatory.  Closes #18529.

abeyad commented on core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java:
```diff
@@ -359,38 +359,32 @@ public BlobStoreIndexShardSnapshot loadSnapshot() {
+                    } catch (IOException e) {
+                        // We cannot delete index file - this is fatal, we cannot continue, otherwise we might end up
+                        // with references to non-existing files
+                        throw new IndexShardSnapshotFailedException(shardId, "error deleting index files during cleanup", e);
```
Since we know the individual file here, we could give a more descriptive error message by saying "error deleting index file [" + blobName + "] during cleanup"

gfyoung replied:
Ah, yes, given the `lambda` format, that certainly makes more sense.  Done.

abeyad commented:
@gfyoung I reviewed the code and it looks good, I put one minor comment.  Also, it would be better to have a more descriptive commit message.  The title for the commit message is good, but there should be a body for the commit message enumerating the two methods that were removed from the interface and why they were removed (i.e. explaining that its a cleaner interface to have just one method, and we gain nothing by the other two methods as they don't afford us atomic deletes anyway).  

Thanks for taking this on!

gfyoung commented:
@abeyad : Certainly.  I've updated the commit message for the duplicate `deleteBlob` method removal, and added a more descriptive error message as requested.  Ready to merge then if there are no other concerns.

abeyad commented:
LGTM

tlrx commented on plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java:
```diff
@@ -85,16 +85,6 @@ public void deleteBlob(String blobName) throws IOException {
-
-    @Override
-    public void deleteBlobs(Collection<String> blobNames) throws IOException {
-        blobStore.deleteBlobs(buildKeys(blobNames));
```
I think that there is now dead code behind this. Can you please verify?

gfyoung replied:
I think I removed all of the dead code that I could possibly remove (see bottom of file).

tlrx replied:
I'm talking about `GoogleCloudStorageBlobStore.deleteBlobs()` method. If it's not used anymore in the `BlobContainer` I think it is useless now.

gfyoung replied:
`deleteBlobs` is called by `deleteBlobsByPrefix` which in turn is called by `delete`.  I suppose I could refactor to put everything in under `delete`?

tlrx commented:
I left a comment.

s1monw commented on core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java:
```diff
@@ -359,38 +359,33 @@ public BlobStoreIndexShardSnapshot loadSnapshot() {
             // delete old index files first
-            for (String blobName : blobs.keySet()) {
-                // delete old file lists
+            blobs.keySet().forEach(blobName -> {
```
can we please stick to the `for` loop. We should be consistent in our loops unless it's really necessary or beneficial

gfyoung replied:
Done.

s1monw commented on core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java:
```diff
@@ -359,38 +359,33 @@ public BlobStoreIndexShardSnapshot loadSnapshot() {
-            // now go over all the blobs, and if they don't exists in a snapshot, delete them
-            for (String blobName : blobs.keySet()) {
+            // now go over all the blobs, and if they don't exist in a snapshot, delete them
+            blobs.keySet().forEach(blobName -> {
```
same here

gfyoung replied:
Done.

s1monw commented:
awesome cleanup - I think we need a note in the migration guide and should deprecate these methods in 2.x @abeyad can you take care of this once it's in?

abeyad commented:
@s1monw will do

jasontedor commented on core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java:
```diff
@@ -359,38 +359,33 @@ public BlobStoreIndexShardSnapshot loadSnapshot() {
+                        try {
+                            blobContainer.deleteBlob(blobName);
+                        } catch (IOException e) {
+                            logger.debug("[{}] [{}] error deleting blob [{}] during cleanup", e, snapshotId, shardId, blobName);
```
Why are we swallowing this? I know that's the way that it was before, but is it right?

abeyad replied:
@imotov could you shed some light on why we swallow the `IOException` here for data files?

I think a future enhancement (not part of this PR) should be to throw the `IOException` and make the callers handle it.  Unless there is a compelling reason not to do it that way.

imotov commented:
-1. As @tlrx said before this will have a significant negative impact on performance of the delete operation on slow repositories. I think we should instead take advantage batch deletes on S3 the same way [as we do it on GCE](https://github.com/elastic/elasticsearch/blob/e7eb664c78dce2451dbeb55db0c0b761a65b295b/plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobStore.java#L250). Snapshot deletion can lead to potential deletion of a large number of files at ones and we will have to wait for roundtrip for each operation before preceding to delete the next file.

s1monw commented:
> -1. As @tlrx said before this will have a significant negative impact on performance of the delete operation on slow repositories. I think we should instead take advantage batch deletes on S3 the same way as we do it on GCE. Snapshot deletion can lead to potential deletion of a large number of files at ones and we will have to wait for roundtrip for each operation before preceding to delete the next file.

what is significant? 1 second per file? 100ms per fiel? I think API simplicity should be preferred compared to batching. The interfaces we use for S/R are extremely polluted with optimizations I really wonder if it's worth it.

imotov commented:
> what is significant? 

I would expect the latency for the delete request to be anywhere between 30ms and 300ms per file depending on location. Assuming, that we delete a snapshot with 100 shards, 20 files per shard with average latency of 50ms per file, we are talking about the difference between 5 sec and 1 minute 40 sec to delete the snapshot.

tlrx commented:
> -1. As @tlrx said before this will have a significant negative impact on performance of the delete operation on slow repositories.

@imotov That's right, I agree with this statement. But what convinced me to remove the multiple variations of `deleteBlob` is that batch deletions make exception and error handling complex. Now we have a Task API I think it is OK to have longer snapshot deletion as long as we can regularly check the progress.

imotov commented:
@tlrx not sure I see how Task API would help with potential 10-20 times slow down in snapshot deletion speed. We can get the performance back by running deletes in parallel but that will increase complexity on other layers and might start triggering throttling on some providers that are not happy with too many small simultaneous requests rushing in. I don't see how this is a good change for S3 and GCE plugins, but if you and @dadoonet like it, I am not going to argue.

dadoonet commented:
I was looking at this today. Actually this `deleteBlobs()` method has been added because of #12697.
If we remove this method, we can close #12697 as it won't be implemented.

We can remove deleteBlobs for now and reimplement if we need to optimize at some point. It will be easier than years ago because plugins and core code lives now in the same repo.

So I'm not against this cleanup.

gfyoung commented:
Okay, so let's quickly summarise what's going on here (easier for myself to understand):

@abeyad proposed the change to simplify the interface <a href="https://github.com/elastic/elasticsearch/issues/18529">here</a> and has given the green light to this PR

@imotov had concerns about latency for S3 and GCE but would not object if @tlrx and @dadoonet both gave the green light to this PR

@tlrx did not believe the latency impact would be too significant and AFAIU has approved this change

@dadoonet also appears to have given the green light for this change AFAIU

So...can we merge this :question:

tlrx commented:
> @tlrx did not believe the latency impact would be too significant and AFAIU has approved this change

Sorry, that's not what I said. I agree with @imotov and I also think that this will have a significant negative impact on performance. On the other hand, Ali, Simon and Robert have the argument that this change make the API cleaner and easier to maintain. I do agree with this too, as well as error handling in batch requests can be tricky to handle correctly too. These points make me think that we can sacrifice some perf at snapshot deletion time in favor of cleaner code if that make our life (and user's life too) better. That's why I'm OK with this change.

gfyoung commented:
@tlrx : updated my comment to clarify based on what you said

My question still stands though.

s1monw commented:
maybe we can add a compromise here and try to interpret a wildcard suffix. If somebody calls `delete("foo*")` we try to fetch all files matching the prefix / wildcard and delete them? This would simplify the interface and we can make batching an impl detail?

gfyoung commented:
@s1monw : integrating `deleteBlobsByPrefix` functionality into `deleteBlobs` sounds like an interesting idea.  Depending on how much traction it gets, I wonder though if it might be best served for a separate PR as follow-up?

abeyad commented:
@gfyoung Sorry I just realized you didn't get a response to this.  Since we are doing the cleanup in this PR, I think it makes more sense to have a deleteBlob with potentially a wildcard prefix string passed in for this PR, because then we won't be getting rid of the bulk updates code, only to reinsert it again in another PR.  Does that make sense?

gfyoung commented:
@abeyad : if you read your explanation again, you didn't really provide justification for doing it now.

On second look, while the idea proposed by @s1monw is interesting, if you look at how `deleteBlobsByPrefix` is implemented in `AbstractBlobContainer`, you can see that it's really just a `for-loop` in most cases that calls `deleteBlob` in the process.  Suddenly AFAICT, integration is not so simple.  

Unless we have functionality that can do `regex` deletions (which I don't think we can do across ALL implementations), essentially we will have to go through some gymnastics of refactoring the current `deleteBlob` code to make sure that it checks whether we have a wildcard and then somehow integrate a for-loop?

Consequently, my thought process is shifting towards just removing `deleteBlobs` but leaving `deleteBlobsByPrefix` unless a nice, more suitable solution can be proposed instead of the one I just described.

s1monw commented:
@imotov will this solve you objection ^^

gfyoung commented:
@imotov: Actually, #18815 (also my PR) got reverted but will be re-merged once a nagging test failure has been fixed.  Not sure if that changes your stance.

IINM, are you no longer objecting to the deletions (i.e. we can even remove the byPrefix method as has already been done)?

gfyoung commented:
@imotov : any response to my comments?

@Everyone : can this be merged if there are no other concerns?

abeyad commented:
Since none of the current blob container implementations take advantage of batching with deleteBlobsByPrefix, the only one that will is @tlrx 's Google implementation.  As discussed with @s1monw, we will merge this PR now and @tlrx can add an optional wildcard suffix to the single `deleteBlob` interface so batching can be done under the hood in working on the Google storage blob container implementation.

abeyad commented:
@gfyoung thanks for your work on this!
